airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigRepository.java

```diff
@@ -627,35 +636,84 @@ public void updateConnectionState(final UUID connectionId, final State state) th
-    final String catalogHash = hashFunction.hashBytes(Jsons.serialize(catalog).getBytes(
+    final String catalogHash = hashFunction.hashBytes(Jsons.serialize(airbyteCatalog).getBytes(
         Charsets.UTF_8)).toString();
-    ActorCatalog actorCatalog = new ActorCatalog()
-        .withCatalog(Jsons.jsonNode(catalog))
-        .withId(UUID.randomUUID())
-        .withCatalogHash(catalogHash);
-    final Optional<ActorCatalog> existingCatalog = findExistingCatalog(actorCatalog);
-    if (existingCatalog.isPresent()) {
-      actorCatalog = existingCatalog.get();
-    } else {
-      persistence.writeConfig(ConfigSchema.ACTOR_CATALOG,
-          actorCatalog.getId().toString(),
-          actorCatalog);
-    }
-    final ActorCatalogFetchEvent actorCatalogFetchEvent = new ActorCatalogFetchEvent()
-        .withActorCatalogId(actorCatalog.getId())
-        .withId(UUID.randomUUID())
-        .withConfigHash(configurationHash)
-        .withConnectorVersion(connectorVersion)
-        .withActorId(sourceId);
-    persistence.writeConfig(ConfigSchema.ACTOR_CATALOG_FETCH_EVENT,
-        actorCatalogFetchEvent.getId().toString(),
-        actorCatalogFetchEvent);
+    final Map<UUID, AirbyteCatalog> catalogs = findCatalogByHash(catalogHash, context);
+
+    for (final Map.Entry<UUID, AirbyteCatalog> entry : catalogs.entrySet()) {
+      if (entry.getValue().equals(airbyteCatalog)) {
+        return entry.getKey();
+      }
+    }
+
+    final UUID catalogId = UUID.randomUUID();
+    context.insertInto(ACTOR_CATALOG)
+        .set(ACTOR_CATALOG.ID, catalogId)
+        .set(ACTOR_CATALOG.CATALOG, JSONB.valueOf(Jsons.serialize(airbyteCatalog)))
+        .set(ACTOR_CATALOG.CATALOG_HASH, catalogHash)
+        .set(ACTOR_CATALOG.CREATED_AT, timestamp)
+        .set(ACTOR_CATALOG.MODIFIED_AT, timestamp).execute();
+    return catalogId;
+  }
+
+  public Optional<AirbyteCatalog> getActorCatalog(final UUID actorId,
+                                                  final String actorVersion,
+                                                  final String configHash)
+      throws IOException {
+    final Result<Record1<JSONB>> records = database.transaction(ctx -> ctx.select(ACTOR_CATALOG.CATALOG)
+        .from(ACTOR_CATALOG).join(ACTOR_CATALOG_FETCH_EVENT)
+        .on(ACTOR_CATALOG.ID.eq(ACTOR_CATALOG_FETCH_EVENT.ACTOR_CATALOG_ID))
+        .where(ACTOR_CATALOG_FETCH_EVENT.ACTOR_ID.eq(actorId))
+        .and(ACTOR_CATALOG_FETCH_EVENT.ACTOR_VERSION.eq(actorVersion))
+        .and(ACTOR_CATALOG_FETCH_EVENT.CONFIG_HASH.eq(configHash))
+        .orderBy(ACTOR_CATALOG_FETCH_EVENT.CREATED_AT.desc()).limit(1)).fetch();
+
+    if (records.size() >= 1) {
+      final JSONB record = records.get(0).get(ACTOR_CATALOG.CATALOG);
+      return Optional.of(Jsons.deserialize(record.toString(), AirbyteCatalog.class));
+    }
+    return Optional.empty();
+
+  }
+
+  public UUID writeActorCatalogFetchEvent(final AirbyteCatalog catalog,
```

Comment:
It would be helpful if these methods had javadocs describing them. The thing that would be useful to explain is how the ActorCatalogFetchEvent and the ActorCatalog tables relate to one another, since it wasn't really obvious to me just from reading the code

Reply:
Done